tests/Doctrine/Tests/ORM/Functional/GH10747Test.php

```diff
@@ -0,0 +1,194 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Doctrine\Tests\ORM\Functional;
+
+use Doctrine\Common\Collections\ArrayCollection;
+use Doctrine\Common\Collections\Collection;
+use Doctrine\DBAL\Platforms\AbstractPlatform;
+use Doctrine\DBAL\Types\Type as DBALType;
+use Doctrine\ORM\Mapping as ORM;
+use Doctrine\ORM\Mapping\Column;
+use Doctrine\ORM\Mapping\Entity;
+use Doctrine\ORM\Mapping\Id;
+use Doctrine\ORM\Mapping\Table;
+use Doctrine\Tests\DbalTypes\CustomIdObject;
+use Doctrine\Tests\OrmFunctionalTestCase;
+
+use function method_exists;
+use function str_replace;
+
+/**
+ * Functional tests for asserting that orphaned children in a OneToMany relationship get removed with a custom identifier
+ *
+ * @group GH10747
+ */
+final class GH10747Test extends OrmFunctionalTestCase
+{
+    protected function setUp(): void
+    {
+        parent::setUp();
+
+        if (! DBALType::hasType(GH10747CustomIdObjectHashType::class)) {
+            DBALType::addType(GH10747CustomIdObjectHashType::class, GH10747CustomIdObjectHashType::class);
+        }
+
+        $this->setUpEntitySchema([GH10747Article::class, GH10747Credit::class]);
+    }
+
+    public function testOrphanedOneToManyDeletesCollection(): void
+    {
+        $object = new GH10747Article(
+            new CustomIdObject('article')
+        );
+
+        $creditOne = new GH10747Credit(
+            $object,
+            'credit1'
+        );
+
+        $creditTwo = new GH10747Credit(
+            $object,
+            'credit2'
+        );
+
+        $object->setCredits(new ArrayCollection([$creditOne, $creditTwo]));
+
+        $this->_em->persist($object);
+        $this->_em->persist($creditOne);
+        $this->_em->persist($creditTwo);
+        $this->_em->flush();
+
+        $id = $object->id;
+
+        $object2 = $this->_em->find(GH10747Article::class, $id);
+
+        $creditThree = new GH10747Credit(
+            $object2,
+            'credit3'
+        );
+
+        $object2->setCredits(new ArrayCollection([$creditThree]));
+
+        $this->_em->persist($object2);
+        $this->_em->persist($creditThree);
+        $this->_em->flush();
+
+        $currentDatabaseCredits = $this->_em->createQueryBuilder()
+            ->select('c.id')
+            ->from(GH10747Credit::class, 'c')
+            ->getQuery()
+            ->execute();
+
+        self::assertCount(1, $currentDatabaseCredits);
+    }
+}
+
+/**
+ * @Entity
+ * @Table
+ */
+class GH10747Article
+{
+    /**
+     * @Id
+     * @Column(type="Doctrine\Tests\ORM\Functional\GH10747CustomIdObjectHashType")
+     * @var CustomIdObject
+     */
+    public $id;
+
+    /**
+     * @ORM\OneToMany(targetEntity="GH10747Credit", mappedBy="article", orphanRemoval=true)
+     *
+     * @var Collection
```

Comment:
Let's be consistent, please.

```suggestion
     * @var Collection<int, GH10747Credit>
```